torch/csrc/jit/interpreter.cpp

```diff
@@ -611,173 +635,239 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
   // minimizing the total number or register
   std::vector<IValue> registers;
 
+  struct Frame {
+    std::shared_ptr<CodeImpl> function;
+    size_t pc;
+  };
+
+  // saved-by-value stuff that can exist on the stack inside runInterpreter
+  struct ActiveFrame {
+    size_t pc;
+    Instruction* instructions;
+    IValue* constants;
+    Operation* operators;
+    std::shared_ptr<script::Function>* functions;
+    ActiveFrame(const Frame& frame)
+        : pc(frame.pc),
+          instructions(frame.function->instructions_.data()),
+          constants(frame.function->constant_table_.data()),
+          operators(frame.function->operator_table_.data()),
+          functions(frame.function->function_table_.data()) {}
+  };
+
+  std::vector<Frame> frames;
+
   c10::intrusive_ptr<InterpreterStateImpl> intrusive_from_this() {
     c10::raw::intrusive_ptr::incref(this);
     return c10::intrusive_ptr<InterpreterStateImpl>::reclaim(this);
   }
 
+  void enterFrame(const Code& code) {
+    frames.emplace_back(Frame{code.pImpl, 0});
+    registers.resize(registers.size() + code.pImpl->register_size_);
+    // frames.back().function->dump(std::cout);
+  }
+
+  void leaveFrame() {
+    registers.resize(registers.size() - frames.back().function->register_size_);
+    frames.pop_back();
+  }
+
+  // relative to the end of the register list so that when we call
+  // functions we are referring to the registers of the currenly executing
+  // function.
+  IValue& reg(size_t reg) {
+    return *(registers.end() - reg);
+  }
+
   bool runImpl(Stack& stack) {
     // if we have never run before, then we might have to return the
     // stack when we suspend, record where it starts so we return the right
     // stack
     if (stack_start_ == -1) {
-      TORCH_INTERNAL_ASSERT(stack.size() >= function->n_inputs);
-      stack_start_ = stack.size() - function->n_inputs;
+      TORCH_INTERNAL_ASSERT(stack.size() >= frames.back().function->n_inputs);
+      stack_start_ = stack.size() - frames.back().function->n_inputs;
     } else {
       // during restarts, all of the stack is always our own, so we leave
       // nothing
       stack_start_ = 0;
     }
+
+    ActiveFrame af(frames.back());
     try {
-      return runInterpreterLoop(stack);
+      while (true) {
```

Comment:
This is just moving the interpreter loop inline here because the exception handling stuff needs to see `af.pc`. The ops themselves, except for CALL have not meaningfully changed.